src/util/graph.h

```diff
@@ -570,6 +579,54 @@ void grapht<N>::make_chordal()
   }
 }
 
+/// Find a topological order of the nodes if graph is DAG, return empty list for
+/// non-DAG or empty graph. Uses Kahn's algorithm running in O(#edges+#nodes).
+template<class N>
+std::list<typename grapht<N>::node_indext> grapht<N>::topsort()
+{
+  std::list<node_indext> nodelist;
+  std::queue<node_indext> indeg0_nodes;
+  std::map<node_indext, size_t> in_deg;
+
+  // abstract graph as in-degree of each node
+  for(node_indext idx=0; idx<nodes.size(); idx++)
+  {
+    if(in(idx).empty())
+      indeg0_nodes.push(idx);
+    in_deg[idx]=in(idx).size();
+  }
+
+  while(!indeg0_nodes.empty())
+  {
+    node_indext source=indeg0_nodes.front();
+    indeg0_nodes.pop();
+    nodelist.push_back(source);
+
+    for(const auto &edge : out(source))
+    {
+      const node_indext target=edge.first;
+      INVARIANT(in_deg[target]!=0, "In-degree of node cannot be 0.");
```

Comment:
Nit picking/question: with `throw` we even have a lint check that the message does not start with an uppercase character; should the same apply to invariants?